experimental/ir/fdp.go

```diff
@@ -0,0 +1,396 @@
+// Copyright 2020-2025 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ir
+
+import (
+	"slices"
+
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/descriptorpb"
+
+	"github.com/bufbuild/protocompile/experimental/ast/predeclared"
+	"github.com/bufbuild/protocompile/experimental/ast/syntax"
+	"github.com/bufbuild/protocompile/experimental/ir/presence"
+	"github.com/bufbuild/protocompile/experimental/seq"
+	"github.com/bufbuild/protocompile/internal/ext/slicesx"
+)
+
+// DescriptorSet generates a FileDescriptorSet for the given files, and returns the
+// result as an encoded byte slice.
+//
+// The resulting FileDescriptorSet is always fully linked: it contains all dependencies except
+// the WKTs, and all names are fully-qualified.
+func DescriptorSet(files []File, options ...DescriptorOption) ([]byte, error) {
+	var dg descGenerator
+	for _, opt := range options {
+		if opt != nil {
+			opt(&dg)
+		}
+	}
+
+	fds := new(descriptorpb.FileDescriptorSet)
+	dg.files(files, fds)
+	return proto.Marshal(fds)
+}
+
+// DescriptorProto generates a single FileDescriptorProto for file, and returns the
+// result as an encoded byte slice.
+//
+// The resulting FileDescriptorProto is fully linked: all names are fully-qualified.
+func DescriptorProto(file File, options ...DescriptorOption) ([]byte, error) {
+	var dg descGenerator
+	for _, opt := range options {
+		if opt != nil {
+			opt(&dg)
+		}
+	}
+
+	fdp := new(descriptorpb.FileDescriptorProto)
+	dg.file(file, fdp)
+	return proto.Marshal(fdp)
+}
+
+// DescriptorOption is an option to pass to [DescriptorSet] or [DescriptorProto].
+type DescriptorOption func(*descGenerator)
+
+// ExcludeFiles is an option that causes all files that match the given
+// predicate to be excluded from the output of [DescriptorSet].
+func ExcludeFiles(p func(File) bool) DescriptorOption {
+	return func(dg *descGenerator) { dg.exclude = append(dg.exclude, p) }
+}
```

Comment:
What is the use case for this again? I know before you had something to omit the well-known types. But I'm trying to think of an actual use case -- things like the LSP or the buf CLI, for example, are going to want everything. If the use case is niche, doesn't seem valuable to make an exported function for it. After all, users can always filter the returned file descriptor set themselves.

Reply:
Let's just drop it now and add it latger if needed.